risc0/zkvm/src/guest/env.rs

```diff
@@ -96,16 +97,41 @@ pub(crate) fn finalize() {
     }
 }
 
-/// Exchange data with the host, returning the data from the host
-/// as a slice of bytes.
-pub fn send_recv(channel: u32, buf: &[u8]) -> &'static [u8] {
-    unsafe { sys_io(channel, buf.as_ptr(), buf.len()) }
+/// Send data to the host.  The host may return two words back.
+pub fn send_host(channel: u32, buf: &[u8]) -> (u32, u32) {
+    unsafe { sys_io(&mut [] as _, 0, buf.as_ptr(), buf.len(), channel) }
 }
 
-/// Exchange data with the host, returning the data from the host as
-/// a slice of words and the length in bytes.
-pub fn send_recv_as_u32(channel: u32, buf: &[u8]) -> &'static [u32] {
-    bytemuck::cast_slice(send_recv(channel, buf))
+/// Exchange data with the host.
+pub fn send_recv_sized(
+    channel: u32,
+    to_host_buf: &[u8],
+    from_host_bytes: usize,
+) -> (u32, u32, &'static [u8]) {
+    unsafe {
+        let nwords = (from_host_bytes + WORD_SIZE - 1) / WORD_SIZE;
+        let from_host_buf = zkvm_abi_alloc_words(nwords);
+        let (a0, a1) = sys_io(
+            from_host_buf,
+            nwords,
+            to_host_buf.as_ptr(),
+            to_host_buf.len(),
+            channel,
+        );
+        (
+            a0,
+            a1,
+            slice::from_raw_parts(from_host_buf.cast(), from_host_bytes),
+        )
+    }
+}
+
+/// Exchange data with host.  The channel should return the length of
+/// the data to read when presented with an empty output buffer.
+pub fn send_recv(channel: u32, to_host_buf: &[u8]) -> &'static [u8] {
+    let (nbytes, _) = send_host(channel, to_host_buf);
+    let (_, _, bytes) = send_recv_sized(channel, to_host_buf, nbytes as usize);
```

Comment:
It seems like if we're going to have this usage pattern, would it make more sense to have 2 different SYS_TX and SYS_RX syscalls rather than combining into SYS_IO?